google/cloud/spanner/result_set.h

```diff
@@ -33,10 +34,182 @@ class ResultSetSource {
+    return RowParser<RowType>(
+        [this]() mutable { return source_->NextValue(); });
+  }
+
+  /**
+   * Retrieve the timestamp at which the read occurred.
+   *
+   * Only available if a read-only transaction was used, and the timestamp
+   * was requested by setting `return_read_timestamp` true.
+   */
+  optional<Timestamp> ReadTimestamp() const {
+    auto metadata = source_->Metadata();
+    if (metadata.has_value() && metadata->has_transaction() &&
+        metadata->transaction().has_read_timestamp()) {
+      return internal::FromProto(metadata->transaction().read_timestamp());
+    }
+    return optional<Timestamp>();
+  }
+
+ private:
+  std::unique_ptr<internal::ResultSetSource> source_;
+};
+
+/**
+ * Represents the result of a query operation using
+ * `spanner::Client::ExecuteQuery()`.
+ *
+ * @note `ExecuteQueryResult` returns both the data for the operation, as
+ * a single-pass, input range returned by `Rows()`, as well as the metadata for
+ * the results, query plan (if requested), and execution statistics
+ * (if requested).
+ */
+class ExecuteQueryResult {
+ public:
+  ExecuteQueryResult() = default;
+  explicit ExecuteQueryResult(std::unique_ptr<internal::ResultSetSource> source)
+      : source_(std::move(source)) {}
+
+  // This class is movable but not copyable.
+  ExecuteQueryResult(ExecuteQueryResult&&) = default;
+  ExecuteQueryResult& operator=(ExecuteQueryResult&&) = default;
+
+  /**
+   * Returns a `RowParser` which can be used to iterate the returned `Row`s.
+   *
+   * Since there is a single result stream for each `ResultSet` instance, users
+   * should not use multiple `RowParser`s from the same `ResultSet` at the same
+   * time. Doing so is not thread safe, and may result in errors or data
+   * corruption.
+   */
+  template <typename RowType>
+  RowParser<RowType> Rows() {
+    return RowParser<RowType>(
+        [this]() mutable { return source_->NextValue(); });
+  }
+
+  /**
+   * Retrieve the timestamp at which the read occurred.
+   *
+   * Only available if a read-only transaction was used, and the timestamp
+   * was requested by setting `return_read_timestamp` true.
+   */
+  optional<Timestamp> ReadTimestamp() const {
+    auto metadata = source_->Metadata();
+    if (metadata.has_value() && metadata->has_transaction() &&
+        metadata->transaction().has_read_timestamp()) {
+      return internal::FromProto(metadata->transaction().read_timestamp());
+    }
+    return optional<Timestamp>();
+  }
+
+  /**
+   * Returns a collection of key value pair statistics for the query execution.
+   *
+   * @note Only available when the query is profiled and all results have been
+   * read.
+   */
+  optional<std::unordered_map<std::string, std::string>> QueryStats() const;
+
+  /**
+   * Returns the plan of execution for the query.
+   *
+   * @note Only available when the query is profiled or when the plan is
+   * explicitly requested.
+   */
+  optional<QueryPlan> QueryExecutionPlan() const;
```

Comment:
Since we have no way to request the query plan at this point, I'd recommend removing this. We can add it back if/when we add support for requesting the query plan. WDYT?

Reply:
created issue #930, will document header indicating these functions are not yet implemented, pending resolve the issue.